homeassistant/components/light/yeelight.py

```diff
@@ -6,157 +6,283 @@
+                  SUPPORT_COLOR_TEMP)
+
+SUPPORTS = (SUPPORT_BRIGHTNESS |
+            SUPPORT_TRANSITION |
+            SUPPORT_FLASH)
+
+
+def _cmd(func):
+    """A wrapper to catch exceptions from the bulb."""
+    def _wrap(self, *args, **kwargs):
+        import yeelight
+        try:
+            _LOGGER.debug("Calling %s with %s %s", func, args, kwargs)
+            return func(self, *args, **kwargs)
+        except yeelight.BulbException as ex:
+            _LOGGER.error("Error when calling %s: %s", func, ex)
+
+    return _wrap
+
 
 def setup_platform(hass, config, add_devices, discovery_info=None):
     """Setup the Yeelight bulbs."""
     lights = []
     if discovery_info is not None:
-        device = {'name': discovery_info['hostname'],
-                  'ipaddr': discovery_info['host']}
-        lights.append(YeelightLight(device))
+        _LOGGER.debug("Adding autodetected %s", discovery_info['hostname'])
+
+        # not using hostname, as it seems to vary.
+        name = "yeelight_%s_%s" % (discovery_info["device_type"],
+                                   discovery_info["properties"]["mac"])
+        device = {'name': name, 'ipaddr': discovery_info['host']}
+
+        default_config = DEVICE_SCHEMA({'transition': DEFAULT_TRANSITION})
+        lights.append(YeelightLight(device, default_config))
     else:
         for ipaddr, device_config in config[CONF_DEVICES].items():
+            _LOGGER.debug("Adding configured %s", device_config[CONF_NAME])
+
             device = {'name': device_config[CONF_NAME], 'ipaddr': ipaddr}
-            lights.append(YeelightLight(device))
+            lights.append(YeelightLight(device, device_config))
 
     add_devices(lights)
 
 
 class YeelightLight(Light):
     """Representation of a Yeelight light."""
 
-    def __init__(self, device):
+    def __init__(self, device, config):
         """Initialize the light."""
-        import pyyeelight
-
+        self.config = config
         self._name = device['name']
         self._ipaddr = device['ipaddr']
-        self.is_valid = True
-        self._bulb = None
-        self._state = None
-        self._bright = None
-        self._rgb = None
-        self._ct = None
-        try:
-            self._bulb = pyyeelight.YeelightBulb(self._ipaddr)
-        except socket.error:
-            self.is_valid = False
-            _LOGGER.error("Failed to connect to bulb %s, %s", self._ipaddr,
-                          self._name)
+
+        self._supported_features = SUPPORTS
+        self.__bulb = None
+        self.__properties = None
+
+    @property
+    def supported_features(self):
+        """Flag supported features."""
+        return self._supported_features
 
     @property
     def unique_id(self):
         """Return the ID of this light."""
         return "{}.{}".format(self.__class__, self._ipaddr)
 
     @property
+    def color_temp(self):
+        """Return the color temperature."""
+        temp_in_k = self._properties.get("ct", None)
```

Comment:
This is just how things are done in hass. Whenever the user changes anything update will be called afterwards. The Entity API is written with the intent that all calculations happen in update, and it is already smart enough to call it when it needs new information. The user won't ever get data directly from the entity, only from the state machine.